GCP RDB Loader

spenes opened the pull request:
This PR contains changes for extracting out AWS related parts from Loader. This is required for our plan to port RDB Loader to GCP. 

Currently, AWS services are directly used without additional abstraction. However, in order to port RDB Loader to GCP, we need to use counterpart GCP services and AWS related bits needs to be extracted out from codebase. With this way, we can easily plug respective GCP services to application.

 The notable changes are following:
* New `cloud` submodule. This module contains interfaces for common interactions with cloud services. Currently, it contains interfaces for blob storage and queue. 
* Removing `aws` submodule. Previously `aws` submodule had been containing entities for interacting with SQS and SNS. They are refactored to adhere new `Queue.Producer` and `Queue.Consumer`  interfaces in `cloud` module. Also, they are moved under `cloud` module.
* S3 -> BlobStorage. Previously `common` submodule had been containing `S3` entity. This entity contained bits necessary to interact with S3 such as constructing S3 paths, splitting these paths etc. `S3` is used in transformer-kinesis as well. In order to make this entity suitable for GCP as well, its name is changed and Google Cloud Storage related bits are added.

TODOs:
- [x] Extract common cloud related parts in Loader and Stream Transformer
- [x] Integrate GCP services (Pubsub, GCS) to Loader
- [x] Extract the cloud related parts from transformer-kinesis and move non cloud related parts to common-transformer-stream module
- [x] Create transformer-pubsub
- [ ] Add example configs for GCP variants
- [ ] Add tests for GCP counterparts of the configs
- [x] Update CI/CD actions for new modules
- [ ] Make Parquet transformation work with GCS
- [x] Make Pubsub configuration values configurable through config file 
- [x] Create secret store for GCP

spenes commented:
@oguzhanunlu raised a few points related with module structure. After some brainstorming, we've decided that it would better to move `cloud` module under `common` module as subpackage. 

Also, we need to decide how new services will be integrated to application. There are three ways as far as I see:

1) Cloud-specific parts can be configured through config file. For example, which queue service will be used by loader can be specified through config file.

2) Creating separate module for every cloud-destination combination. e.g. aws-databricks-loader, gcp-databricks-databricks etc. 

3) Creating cloud specific binding modules. Creating artifacts for every cloud-destination combination can be handled on SBT side with this way.

I personally vote for option 1. The advantage of it is simplicity. The disadvantage is that it will require some if-else. However, this part can be fairly isolated therefore it shouldn't pollute all the parts. 

I know this isn't way we follow in other applications like enrich or collector however unlike those applications, destination part also variable in here and creating separate artifact/module for every cloud-destination combination leads to artifact explosion. 

We might use option 3 to make number of project submodule manageable, however it will be quite experimental setup. If we decide to follow that approach, it might be nice to get a few pointers.

Can I take your opinions on this ? /cc @peel @oguzhanunlu

oguzhanunlu commented:
Not sure if it is option 1 or 3 but I also don't like option 2 -- where I don't want to imagine what happens when we introduce rabbitmq/kafka or firebolt support.

Like Enes, I can imagine what option 1 looks like so that's what I'd do. At the same time, I'm curious to learn more about how option 3 would look in practice -- smells like slf4j bindings

peel commented:
I think sticking to option 1 makes sense from predictability point of view. If we ever decided to extract cloud-specific adapters it would be easier to go from if-else to adapters/bindings than the other way around.

spenes commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/cloud/JsonPathDiscovery.scala:
```diff
@@ -0,0 +1,110 @@
+  def discoverJsonPath(jsonpathAssets: Option[BlobStorage.Folder], shreddedType: Info): DiscoveryAction[F, BlobStorage.Key]
+}
+
+object JsonPathDiscovery {
```
I've placed `JsonPathDiscovery` and `LoadAuthService` under `loader` module since they are only used in the loader. It doesn't look good to have `cloud` packages in both `common` and `loader` modules but this seems best way to me. Let me know if you have other opinions.

spenes commented:
Hey @peel and @oguzhanunlu! I've updated the PR with the changes on Stream Transformer. It would be great if I can get a review on this one. Here is summary of the changes:

Basically, I extracted cloud related parts from `transformer-kinesis` and move cloud agnostic parts to `common-transformer-stream` module. Then, I've created cloud specific modules: `transformer-kinesis` and `transformer-pubsub`

I've continued to use this PR for the changes on Stream Transformer because the common parts for loader and transformer needs to be changed during the changes on Stream Transformer. Therefore, I thought it would be better to continue to use this PR. 

The PR become quite huge however most of the diffs are due to moving files around. Actual work has been done on the `cloud` subpackage under `common` module and new transformer modules (`transformer-pubsub` and `transformer-kinesis`).

voropaevp commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/Transaction.scala:
```diff
@@ -91,7 +89,7 @@ object Transaction {
           Resource.pure[F, String](text)
         case StorageTarget.PasswordConfig.EncryptedKey(StorageTarget.EncryptedConfig(key)) =>
-          Resource.eval(AWS[F].getEc2Property(key.parameterName).map(b => new String(b)))
+          Resource.eval(AWS.getEc2Property[F](key.parameterName).map(b => new String(b)))
```
Implicit dependency between AWS and Redshift with EncryptedKey.  
It should be validated during Config parsing.

spenes replied:
I've created `SecretStore` interface for this purpose. GCP version will use [GCP Secret Manager](https://cloud.google.com/secret-manager) which is the counterpart of EC2 Parameter Store.

voropaevp commented on modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/utils/SSH.scala:
```diff
@@ -53,11 +53,11 @@ object SSH {
       .bastion
       .key
-      .map(_.ec2ParameterStore.parameterName).traverse(AWS[F].getEc2Property)
+      .map(_.ec2ParameterStore.parameterName).traverse(AWS.getEc2Property[F])
```
Similar implicit dependency between SSH and AWS. I think it should be validated in config.

spenes replied:
https://github.com/snowplow/snowplow-rdb-loader/pull/1024#discussion_r987793338

voropaevp commented on modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/cloud/aws/AWS.scala:
```diff
@@ -0,0 +1,132 @@
+// This project
+import com.snowplowanalytics.snowplow.rdbloader.common.cloud.{BlobStorage, Queue}
+
+object AWS {
```
It might benefit from organizing functions a bit further. i.e. object AWS { object AwsQueue{}, AwsAuth{}, etc.}. Same for GCP.scala

voropaevp commented on modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/cloud/BlobStorage.scala:
```diff
@@ -0,0 +1,218 @@
+  def apply[F[_]](implicit ev: BlobStorage[F]): BlobStorage[F] = ev
+
+  /**
+   * Refined type for AWS blob storage bucket, allowing only valid blob storage paths
```
AWS is mentioned in few comment in this file. It should be removed.

voropaevp commented on modules/common/src/test/scala/com.snowplowanalytics.snowplow.rdbloader.common/S3Spec.scala:
```diff
@@ -12,24 +12,23 @@
-      S3.Folder.parse(folder) must beRight
+      BlobStorage.Folder.parse(folder) must beRight
     }
     "support s3n:// prefix" >> {
```
add gs: spec

voropaevp commented on modules/snowflake-loader/src/main/scala/com/snowplowanalytics/snowplow/loader/snowflake/Snowflake.scala:
```diff
@@ -303,7 +303,7 @@ object Snowflake {
         Fragment.const0(s"CREDENTIALS = (AWS_KEY_ID = '${awsAccessKey}' AWS_SECRET_KEY = '${awsSecretKey}' AWS_TOKEN = '${awsSessionToken}')")
```
Would this work while running in GCP?

voropaevp reviewed (changes requested):
Good job on abstracting out the blob storage. 

AWS authenication is leaking in few places, where it used for warehouse authentiction.  Perhaps, there is a place for some unified auth abstracion that would consolidate warehouses.

peel reviewed (commented):
Great job @spenes! Took me a while to get familiar with RDB loader and what's been around and what's new. Looks really well. I think to @voropaevp's point there's a couple of places where we're leaking AWS but not in a major way. The interfaces are coherent and solid, which was arguably the most important bit here.

oguzhanunlu commented on .github/workflows/ci.yml:
```diff
@@ -191,6 +192,10 @@ jobs:
+    - name: Build transformer-pubsub jar
+      env:
+        SKIP_TEST: true
+      run: sbt 'project transformerPubsub' assembly
```
We could replace L175-198 with a single step

```
sbt \
'project redshiftLoader; assembly' \
'project snowflakeLoader; assembly' \
'project databricksLoader; assembly' \
'project transformerBatch; assembly' \
'project transformerKinesis; assembly' \
'project transformerPubsub; assembly'
```

assuming that we don't want simpler `sbt assembly` not to wait on building all projects' fatjars

oguzhanunlu commented on build.sbt:
```diff
@@ -34,15 +65,15 @@ lazy val loader = project
   .settings(libraryDependencies ++= Dependencies.loaderDependencies)
   .settings(excludeDependencies ++= Dependencies.exclusions)
-  .dependsOn(common % "compile->compile;test->test", aws)
+  .dependsOn(common % "compile->compile;test->test", aws % "compile->compile;test->test", gcp % "compile->compile;test->test")
```
`common % "compile->compile;test->test",` can be removed

oguzhanunlu commented on build.sbt:
```diff
@@ -34,15 +65,15 @@ lazy val loader = project
   .settings(addCompilerPlugin(Dependencies.betterMonadicFor))
   .settings(libraryDependencies ++= Dependencies.redshiftDependencies)
-  .dependsOn(common % "compile->compile;test->test", aws, loader % "compile->compile;test->test")
+  .dependsOn(common % "compile->compile;test->test", loader % "compile->compile;test->test")
```
`common % "compile->compile;test->test",` can be removed here and in `redshiftLoaderDistroless` as well

oguzhanunlu commented on build.sbt:
```diff
@@ -51,15 +82,15 @@ lazy val redshiftLoaderDistroless = project
   .settings(addCompilerPlugin(Dependencies.betterMonadicFor))
   .settings(libraryDependencies ++= Dependencies.snowflakeDependencies)
-  .dependsOn(common % "compile->compile;test->test", aws, loader % "compile->compile;test->test")
+  .dependsOn(common % "compile->compile;test->test",loader % "compile->compile;test->test")
```
same point as https://github.com/snowplow/snowplow-rdb-loader/pull/1024/files#r989998181 , `snowflakeLoaderDistroless` as well

oguzhanunlu commented on build.sbt:
```diff
@@ -68,22 +99,22 @@ lazy val snowflakeLoaderDistroless = project
   .settings(BuildSettings.databricksBuildSettings)
   .settings(addCompilerPlugin(Dependencies.betterMonadicFor))
-  .dependsOn(common % "compile->compile;test->test", aws, loader % "compile->compile;test->test")
+  .dependsOn(common % "compile->compile;test->test",loader % "compile->compile;test->test")
```
same point as https://github.com/snowplow/snowplow-rdb-loader/pull/1024/files#r989998181 , `databricksLoaderDistroless` as well
